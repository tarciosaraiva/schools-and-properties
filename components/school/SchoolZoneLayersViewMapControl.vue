<template>
  <div class="layer-container">
    <button title="Display school zones control" @click="toggleControl">
      <font-awesome-icon icon="layer-group" size="lg" />
    </button>
    <div class="layer-options" :class="collapsed ? '' : 'expanded'">
      <label class="checkbox">
        <input type="checkbox" value="primary" @click="handleChecked"> Primary
      </label>
      <label class="checkbox">
        <input type="checkbox" value="p7" @click="handleChecked"> Secondary Yr 7
      </label>
      <label class="checkbox">
        <input type="checkbox" value="p8" @click="handleChecked"> Secondary Yr 8
      </label>
      <label class="checkbox">
        <input type="checkbox" value="p9" @click="handleChecked"> Secondary Yr 9
      </label>
      <label class="checkbox">
        <input type="checkbox" value="p10" @click="handleChecked"> Secondary Yr 10
      </label>
      <label class="checkbox">
        <input type="checkbox" value="p11" @click="handleChecked"> Secondary Yr 11
      </label>
      <label class="checkbox">
        <input type="checkbox" value="p12" @click="handleChecked"> Secondary Yr 12
      </label>
    </div>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'

export default Vue.extend({
  props: {
    toggleLayer: Function
  },
  data () {
    return {
      collapsed: true
    }
  },
  mounted() {
    document.addEventListener('click', this.handleClickOutside)
  },
  destroyed() {
    document.removeEventListener('click', this.handleClickOutside)
  },
  methods: {
    handleClickOutside(event: any) {
      if (!this.$el.contains(event.target)) {
        this.collapsed = true;
      }
    },
    handleChecked (e: any) {
      this.toggleLayer(e.target.value)
    },
    toggleControl () {
      this.collapsed = !this.collapsed
    }
  }
})
</script>

<style scoped>
.layer-container {
  position: relative;
}

.layer-container .layer-options {
  background: rgba(255, 255, 255, 0.85);
  border-radius: 4px;
  border: 0;
  height: 0;
  left: 40px;
  max-height: 260px;
  overflow-x: hidden;
  overflow-y: auto;
  padding: 0;
  position: absolute;
  top: 0;
  transition: all 0.3s ease;
  width: 0;
}

.layer-container .layer-options.expanded {
  border: 1px solid #ccc;
  padding: 0.75rem;
  height: auto;
  width: 9rem;
}

.layer-container .layer-options .checkbox {
  padding: 0.25rem 0;
  font-size: 120%;
  white-space: nowrap;
}

.layer-container .layer-options .checkbox:first-child {
  padding-top: 0;
}

.layer-container .layer-options .checkbox:last-child {
  padding-bottom: 0;
  margin-bottom: 0;
}
</style>
