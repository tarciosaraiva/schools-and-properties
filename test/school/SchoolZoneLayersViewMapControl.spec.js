import { mount, createLocalVue } from '@vue/test-utils'
import { library } from '@fortawesome/fontawesome-svg-core'
import { fas } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'

import SchoolZoneLayersViewMapControl from '~/components/school/SchoolZoneLayersViewMapControl.vue'

library.add(fas)

const localVue = createLocalVue()
localVue.component('FontAwesomeIcon', FontAwesomeIcon)

describe('SchoolZoneLayersViewMapControl', () => {
  const propsData = {
    toggleLayer: jest.fn()
  }

  let wrapper

  beforeEach(() => {
    wrapper = mount(SchoolZoneLayersViewMapControl, { propsData, localVue })
  })

  test('is a Vue instance', () => {
    expect(wrapper.vm).toBeTruthy()
  })

  test('renders correctly with props', () => {
    expect(wrapper).toMatchSnapshot()
  })

  test('should toggle zone control when click on button', async () => {
    const button = wrapper.find('button')
    await button.trigger('click')
    expect(wrapper.vm.collapsed).toBeFalsy()
    // click again to collapse
    await button.trigger('click')
    expect(wrapper.vm.collapsed).toBeTruthy()
  })

  test('should invoke "toggleLayer" when checkbox is clicked', async () => {
    const input = wrapper.findAll('input')
    await input.at(0).trigger('click')
    expect(propsData.toggleLayer).toHaveBeenCalledWith('primary')
  })
})
