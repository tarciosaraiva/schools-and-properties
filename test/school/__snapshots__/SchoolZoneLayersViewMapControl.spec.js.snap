// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SchoolZoneLayersViewMapControl renders correctly with props 1`] = `
<div class="layer-container">
  <button>
    <svg aria-labelledby="svg-inline--fa-title-wO7uG5OPTKqa" class="svg-inline--fa fa-layer-group fa-w-16 fa-lg" data-icon="layer-group" data-prefix="fas" role="img" viewBox="0 0 512 512" xmlns="http://www.w3.org/2000/svg">
      <title class="" id="svg-inline--fa-title-wO7uG5OPTKqa">Display school zones control</title>
      <path class="" d="M12.41 148.02l232.94 105.67c6.8 3.09 14.49 3.09 21.29 0l232.94-105.67c16.55-7.51 16.55-32.52 0-40.03L266.65 2.31a25.607 25.607 0 0 0-21.29 0L12.41 107.98c-16.55 7.51-16.55 32.53 0 40.04zm487.18 88.28l-58.09-26.33-161.64 73.27c-7.56 3.43-15.59 5.17-23.86 5.17s-16.29-1.74-23.86-5.17L70.51 209.97l-58.1 26.33c-16.55 7.5-16.55 32.5 0 40l232.94 105.59c6.8 3.08 14.49 3.08 21.29 0L499.59 276.3c16.55-7.5 16.55-32.5 0-40zm0 127.8l-57.87-26.23-161.86 73.37c-7.56 3.43-15.59 5.17-23.86 5.17s-16.29-1.74-23.86-5.17L70.29 337.87 12.41 364.1c-16.55 7.5-16.55 32.5 0 40l232.94 105.59c6.8 3.08 14.49 3.08 21.29 0L499.59 404.1c16.55-7.5 16.55-32.5 0-40z" fill="currentColor" />
    </svg>
  </button>
  <div class="layer-options">
    <label class="checkbox">
      <input type="checkbox" value="primary"> Primary
    </label>
    <label class="checkbox">
      <input type="checkbox" value="p7"> Secondary Yr 7
    </label>
    <label class="checkbox">
      <input type="checkbox" value="p8"> Secondary Yr 8
    </label>
    <label class="checkbox">
      <input type="checkbox" value="p9"> Secondary Yr 9
    </label>
    <label class="checkbox">
      <input type="checkbox" value="p10"> Secondary Yr 10
    </label>
    <label class="checkbox">
      <input type="checkbox" value="p11"> Secondary Yr 11
    </label>
    <label class="checkbox">
      <input type="checkbox" value="p12"> Secondary Yr 12
    </label>
  </div>
</div>
`;
